pkg/providers/vsphere/virtualmachine/backup.go

```diff
@@ -53,22 +56,40 @@ func BackupVirtualMachine(opts BackupVirtualMachineOptions) error {
 		opts.VMCtx.Logger.Error(err, "failed to get VM properties for backup")
 		return err
 	}
-
 	curExCfg := pkgutil.OptionValues(moVM.Config.ExtraConfig)
 	var ecToUpdate pkgutil.OptionValues
 
-	vmYAML, err := getDesiredVMResourceYAMLForBackup(opts.VMCtx.VM, curExCfg)
-	if err != nil {
-		opts.VMCtx.Logger.Error(err, "failed to get VM resource yaml for backup")
-		return err
-	}
-	if vmYAML == "" {
-		opts.VMCtx.Logger.V(4).Info("Skipping VM resource yaml backup as unchanged")
-	} else {
-		ecToUpdate = append(ecToUpdate, &vimtypes.OptionValue{
-			Key:   vmopv1.VMResourceYAMLExtraConfigKey,
-			Value: vmYAML,
-		})
+	/*
+	 * When VMIncrementalRestore FSS is enabled, perform backups when
+	 * - the backup version annotation matches the current backup version in extraConfig.
+	 * - no backup version annotation is present.
+	 * - the backup version annotation is older than the extraConfig backup version key. This can only happen when reconfigure succeeds to add a new
+	 *   extraConfig backup version but the next reconcile failed to update the annotation on the VM resource.
+	 */
+	if pkgcfg.FromContext(opts.VMCtx).Features.VMIncrementalRestore {
+		if backupVersionAnnotation, ok := opts.VMCtx.VM.Annotations[vmopv1.VirtualMachineBackupVersionAnnotation]; ok && backupVersionAnnotation != "" {
+			currBackupVersionEcKey, _ := curExCfg.GetString(vmopv1.BackupVersionKey)
+			if currBackupVersionEcKey != "" && currBackupVersionEcKey != backupVersionAnnotation {
+				storedVersion, err := strconv.ParseInt(currBackupVersionEcKey, 10, 64)
+				if err != nil {
+					opts.VMCtx.Logger.Error(err, fmt.Sprintf("failed to parse backup version extraConfig key to int64 : %v", currBackupVersionEcKey))
+					return err
+				}
+
+				currVersion, err := strconv.ParseInt(backupVersionAnnotation, 10, 64)
+				if err != nil {
+					opts.VMCtx.Logger.Error(err, fmt.Sprintf("failed to parse backup version annotation to int64: %v", backupVersionAnnotation))
+					return err
+				}
+
+				// if the storedVersion in the backup is older than the annotation, this is a restore.
+				if currVersion > storedVersion {
+					opts.VMCtx.Logger.V(4).Info("Skipping VM backup as a restore is detected , wait for VM registration")
+
+					return fmt.Errorf("backup version drift detected, annotation: %s, backup extraconfig: %s", backupVersionAnnotation, currBackupVersionEcKey)
```

Comment:
Do we need to return an error to force a reconciliation here? After VM registration, VMOP will receive reconciliation requests regardless, so we could just log with higher visibility and return.

Reply:
thanks. I wanted your (or) @aruneshpa's opinion on this. I initially had an implementation that returned nil but changed it to force reconciliation. I'll change it back.